src/Token/DataSet.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * This file is part of Lcobucci\JWT, a simple library to handle JWT and JWS
+ *
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ */
+
+declare(strict_types=1);
+
+namespace Lcobucci\JWT\Token;
+
+/**
+ * @author Luís Otávio Cobucci Oblonczyk <lcobucci@gmail.com>
+ *
+ * @since 4.0.0
+ */
+final class DataSet
+{
+    /**
+     * @var array
+     */
+    private $data;
+
+    /**
+     * @var string
+     */
+    private $encoded;
+
+    public function __construct(array $data, string $encoded)
+    {
+        $this->data = $data;
```

Comment:
The intent here is to encapsulate both values in a common structure for headers and claims (and reduce the repeated methods in the `Token`)...

It simplified things a lot IMO but it's indeed a parallel data structure and there's no way to ensure that `$encoded` is the string representation of the `$data` and of course the validity but I'd say it's an acceptable trade off.

Do you have any suggestion here?